zigpy/device.py

```diff
@@ -139,6 +139,8 @@ async def _initialize(self):
                 continue
             try:
                 await ep.initialize()
+            except asyncio.CancelledError:
+                raise
```

Comment:
should it just log a warning and `break` instead like a regular exception? I don't think we directly cancel `endpoint.initialize()` but we may cancel `device.initialze()` should we guard there instead?

Reply:
Hmm. I think there is a difference between `ep.initialize()` failing (i.e. when the device falls asleep and the radio library throws an exception due to the message being undeliverable) and `device.cancel_initialization()` being called, cancelling the initialization task. The former is common, but the latter seems like it shouldn't be swallowed and then `self.application.remove` be awaited, potentially never cancelling the task if that coroutine takes a long time.

Also, the device object removing itself from its `ControllerApplication` may not be a good idea?? It seems like in this specific case, it would be better for the application to schedule a device initialization in a dedicated task and then remove the device if the initialization fails, instead of the device having a method that does all of this on its own? The `device_init_failure` event is not used by ZHA or anything else and seems like a substitute for just letting an exception happen.

Reply:
I guess my question is: while `device.initialize()` is  awaiting on `ep.initialize()` and `dev.initialize()` is cancelled, is ep.initialize() cancelled too or is it allowed to finish?

But instead, maybe instead of raising and catching a wide exception, we should catch `ZigbeeException` and `asyncio.Timeout` exceptions allowing anything else rise? and on failures, instead of throwing a `Exception` we should raise a more specific `ControllerException` ?

Reply:
Hmm. I expect a task to immediately get cancelled when it's run unless it's doing something extremely critical when cleaning up.

By catching `ZigbeeException` and `asyncio.Timeout`, do you mean doing what currently happens, which is logging a warning and returning? To be honest, I think all of these `try` blocks that swallow exceptions and log them instead should be removed entirely. If whatever is running `get_node_descriptor` or `initialize` wants to itself swallow the exceptions, it should. Otherwise, let the coroutines work normally and leave the higher-level application controlling zigpy to deal with the errors however it wants.